src/client/client_impl.cc

```diff
@@ -79,6 +82,189 @@ google::protobuf::util::Status ClientImpl::CreateFile(
+    // We start reading from the beginning of this chunk, where the offset
+    // equals to chunk_index * common::bytes_per_mb
+    read_file_chunk_request.set_offset_start(chunk_index * 
+                                                 common::bytes_per_mb);
+    read_file_chunk_request.set_length(nbytes);
+    // Access the client-end-point for contacting the chunk server
+    auto server_address(location.server_hostname()+":" + 
+                            std::to_string(location.server_port()));
+    // Prepare the client context 
+    grpc::ClientContext client_context;  
+    common::SetClientContextDeadline(client_context, config_manager_); 
+    // Issue ReadFileChunkRequest and check status
+    StatusOr<ReadFileChunkReply> read_file_chunk_reply_or(
+        chunk_server_service_client_[server_address]->SendRequest(
+            read_file_chunk_request, client_context));
+    
+    // Handle grpc error, log and continue to try the next chunk server
+    if (!read_file_chunk_reply_or.ok()) {
+      LOG(ERROR) << "Read file chunk " << chunk_handle << " from "
+                 << server_address << " failed due to " 
+                 << read_file_chunk_reply_or.status().error_message();
+      continue;
+    }
+
+    // Handle chunk status error, log and continue
+    auto read_file_chunk_reply(read_file_chunk_reply_or.ValueOrDie());
+    switch (read_file_chunk_reply.status()) {
+      case ReadFileChunkReply::UNKNOWN:
+        LOG(ERROR) << "Unknown error while reading " + chunk_handle;
+        continue; 
+      case ReadFileChunkReply::FAILED_NOT_FOUND:
+        LOG(ERROR) << "Chunk not found: " + chunk_handle;
+        continue;
+      case ReadFileChunkReply::FAILED_STALE_VERSION:
+        LOG(ERROR) << "Reading stale version " << std::to_string(chunk_version)
+                   << " for chunk " + chunk_handle;
+        continue;
+      default: break;
+    }
+
+    // This read has been successful, return here. We expect the compiler to
+    // execute copy elision here to "move" the read reply. (It is tempting 
+    // to explicitly write std::move(read_file_chunk_reply) but that is not 
+    // the right approach to do so). 
+    return read_file_chunk_reply;
+  }
+
+  // Failed to read from all chunk servers
+  return google::protobuf::util::Status(
+             google::protobuf::util::error::INTERNAL,
+             "Failed to read from all chunk servers for " + chunk_handle);
+}
+
+google::protobuf::util::StatusOr<std::pair<size_t, void*>> 
+    ClientImpl::ReadFile(const char* filename, size_t offset, size_t nbytes) {
+  // Record the number of bytes that we already read
+  size_t bytes_read(0);
+  // Record the number of bytes to read
+  size_t remain_bytes(nbytes);
+
+  // We simply allocate the necessary size for the buffer. There is a 
+  // possibility that we may not read up to nbytes, but it is ok to pass
+  // this optimization opportunity as it may cost more to realloc
+  void* buffer(malloc(nbytes));
```

Comment:
kind of worried about memory leaks, as we don't know how client will use this (and delete it as necessary). we should def take it into account in the test and other places;

Reply:
The setting for now is the caller (client)'s responsibility to free the resource. (I have fixed the leak in test). We can improve this later if time allows.